driver/fvGFS/fv_ufs_restart_io.F90

```diff
@@ -0,0 +1,544 @@
+
+ contains
+
+ subroutine fv_dyn_restart_register (Atm)
+
+   ! this subroutine must allocate all data buffers and set the variable names
+   ! for restart bundles
+   ! must be consistent with fv_io_register_restart in atmos_cubed_sphere/tools/fv_io.F90
+
+   implicit none
+
+   type(fv_atmos_type), intent(inout) :: Atm
+
+   integer :: isc, iec, jsc, jec, nx, ny, nx_c, ny_c, nx_s, ny_s, nx_e, ny_e
+   integer :: nlon, nlat, mlon, mlat
+   integer :: num, nt, n
+   character(len=64) :: tracer_name
+
+   ! if(mpp_pe() == mpp_root_pe()) write(0,*) ' in fv_dyn_restart_register'
+
+   isc = Atm%bd%isc
+   iec = Atm%bd%iec
+   jsc = Atm%bd%jsc
+   jec = Atm%bd%jec
+   nx  = (iec - isc + 1)
+   ny  = (jec - jsc + 1)
+
+   ! core
+   ! 'u', 'v'
+   nvar3d_core_center = 0
+   if (.not.atm%flagstruct%hydrostatic) then
+      if (atm%flagstruct%make_nh) then ! hydrostatic restarts dont have these variables
+           nvar3d_core_center = nvar3d_core_center + 2 ! 'w', 'dz'
+           if ( atm%flagstruct%hybrid_z ) then
+               nvar3d_core_center = nvar3d_core_center + 1 ! 'ze0'
+           endif
+      else !the restart file has the non-hydrostatic variables
+           nvar3d_core_center = nvar3d_core_center + 2 ! 'w', 'dz'
+           if ( atm%flagstruct%hybrid_z ) then
+               nvar3d_core_center = nvar3d_core_center + 1 ! 'ze0'
+           endif
+      endif
+   endif
+   nvar3d_core_center = nvar3d_core_center + 2 ! 't', 'delp'
+   !--- include agrid winds in restarts for use in data assimilation
+   if (atm%flagstruct%agrid_vel_rst) then
+      nvar3d_core_center = nvar3d_core_center + 2 ! 'ua', 'va'
+   endif
+   core_zsize = size(Atm%u,3)
+
+   call atmosphere_resolution (mlon, mlat, global=.true.)
+   nx_c = nx
+   ny_c = ny
+   if (iec == mlon) then
+      ! we are on task at the 'east' edge of the cubed sphere face or regional domain
+      ! corner arrays should have one extra element in 'i' direction
+      nx_c = nx + 1
+   end if
+   if (jec == mlat) then
+      ! we are on task at the 'north' edge of the cubed sphere face or regional domain
+      ! corner arrays should have one extra element in 'j' direction
+      ny_c = ny + 1
+   end if
+
+   allocate(core_north_var3  (nx  , ny_c,core_zsize,1))
+   allocate(core_east_var3   (nx_c, ny  ,core_zsize,1))
+   allocate(core_center_var3 (nx  , ny  ,core_zsize,nvar3d_core_center))
+   allocate(core_center_var2 (nx  , ny  ,1))
+
+   core_north_var3  = 0.0
+   core_east_var3   = 0.0
+   core_center_var3 = 0.0
+   core_center_var2 = 0.0
+
+   allocate(core_north_var3_names  (1))
+   allocate(core_east_var3_names   (1))
+   allocate(core_center_var3_names (nvar3d_core_center))
+   allocate(core_center_var2_names (1))
+
+   core_north_var3_names(1) = 'u'
+   core_east_var3_names(1) = 'v'
+   n = 1
+   if (.not.atm%flagstruct%hydrostatic) then
+      if (atm%flagstruct%make_nh) then ! hydrostatic restarts dont have these variables
```

Comment:
I am not sure if I understand correctly. But the lines 123-127 look the same as line 129-133, do we need this if statement?